arangod/Cluster/AgencyCache.cpp

```diff
@@ -470,3 +540,114 @@ void AgencyCache::invokeAllCallbacks() const {
   }
   invokeCallbacks(toCall);
 }
+
+
+//TODO Snapshot case to do
+AgencyCache::change_set_t const AgencyCache::changedSince(
+  std::string const& what, consensus::index_t const& last) const {
+
+  static std::vector<std::string> const planGoodies ({
+      AgencyCommHelper::path(PLAN_ANALYZERS) + "/",
+      AgencyCommHelper::path(PLAN_COLLECTIONS) + "/",
+      AgencyCommHelper::path(PLAN_DATABASES) + "/",
+      AgencyCommHelper::path(PLAN_VIEWS) + "/"});
+  static std::vector<std::string> const currentGoodies ({
+      AgencyCommHelper::path(CURRENT_COLLECTIONS) + "/",
+      AgencyCommHelper::path(CURRENT_DATABASES) + "/"});
+  
+  bool get_rest = false;
+  std::unordered_map<std::string, query_t> db_res;
+  query_t rest_res = nullptr;
+
+  decltype(_planChanges) const& changes = (what == PLAN) ? _planChanges : _currentChanges;
+  std::vector<std::string> const& goodies = (what == PLAN) ? planGoodies : currentGoodies;
+  
+  std::unordered_set<std::string> databases;
+  std::shared_lock g(_storeLock);
+  
+  auto tmp = _readDB.nodePtr()->hasAsUInt(AgencyCommHelper::path(what) + "/" + VERSION);
+  uint64_t version = tmp.second ? tmp.first : 0;
+
+  if (last < _lastSnapshot) {
+    get_rest = true;
+    auto keys = _readDB.nodePtr(AgencyCommHelper::path(what) + "/" + DATABASES)->keys();
+    for (auto const& i : keys) {
+      databases.emplace(i);
+    }
+  } else {
+    auto it = changes.lower_bound(last+1); 
+    if (it != changes.end()) {
+      for (; it != changes.end(); ++it) {
+        if (it->second.empty()) { // Need to get rest
+          get_rest = true;
+        }
+        databases.emplace(it->second);
+      }
+      LOG_TOPIC("d5743", TRACE, Logger::CLUSTER)
+        << "collecting " << databases << " from agency cache";
+    } else {
+      LOG_TOPIC("d5734", DEBUG, Logger::CLUSTER) << "no changed databases since " << last;
+      return change_set_t(_commitIndex, version, std::move(db_res), std::move(rest_res));
+    }
+  }
+  
+  if (databases.empty()) {
+    return change_set_t(_commitIndex, version, std::move(db_res), std::move(rest_res));
+  }
+
+  auto query = std::make_shared<arangodb::velocypack::Builder>();
+  {
+    for (auto const& i : databases) {
+      if (!i.empty()) { // database
+        auto query = std::make_shared<arangodb::velocypack::Builder>();
```

Comment:
this redeclares `query` in the inner loop, but it is already declared before this loop.
```suggestion
        query->clear();
```